domain/permission/state/state.go

```diff
@@ -0,0 +1,362 @@
+	targetExistsStmt, err := s.targetExistsStmt()
+	if err != nil {
+		return userAccess, errors.Trace(err)
+	}
+
+	// Insert a permission doc with
+	// * permissionObjectAccess as permission_type_id
+	// * uuid of the user (spec.User) as grant_to
+	// * spec.Target.Key as grant_on
+	newPermission := `
+INSERT INTO permission (uuid, permission_type_id, grant_on, grant_to)
+VALUES ($M.uuid, $M.access_type_id, $M.grant_on, $M.grant_to)
+`
+	insertPermissionStmt, err := sqlair.Prepare(newPermission, sqlair.M{})
+	if err != nil {
+		return userAccess, errors.Trace(err)
+	}
+
+	m := make(sqlair.M, 6)
+	m["grant_on"] = spec.Target.Key
+	m["access_type"] = spec.Access
+	m["object_type"] = spec.Target.ObjectType
+	m["uuid"] = newPermissionUUID.String()
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var result User
+		err = tx.Query(ctx, selectUserStmt, sqlair.M{"name": spec.User}).Get(&result)
+		if err != nil && errors.Is(err, sql.ErrNoRows) {
+			return errors.Annotatef(usererrors.NotFound, "%q", spec.User)
+		} else if err != nil {
+			return errors.Annotatef(err, "getting user with name %q", spec.User)
+		}
+		if result.Disabled {
+			return errors.Annotatef(usererrors.AuthenticationDisabled, "%q", spec.User)
+		}
+		userAccess = result.toCoreUserAccess()
+		m["grant_to"] = result.UUID
+
+		var resultM = sqlair.M{}
+		err = tx.Query(ctx, objectAccessIDStmt, m).Get(&resultM)
+		if err != nil && errors.Is(err, sql.ErrNoRows) {
+			return errors.Annotatef(err, "mismatch in %q, %q", spec.Access, spec.Target.ObjectType)
+		} else if err != nil {
+			return errors.Annotatef(err, "getting id for pair %q, %q", spec.Access, spec.Target.ObjectType)
+		}
+		m["access_type_id"] = resultM["access_type_id"]
+
+		if spec.Target.Key != coredatabase.ControllerNS {
+			// Check that 1 row exists if the grant_on value is not ControllerNS.
+			// The behavior for this check is changing in sqlair, trying to
+			// account for either behavior. Old behavior: return no error and
+			// a slice length of zero. New behavior: if the length is 0, return
+			// ErrNoRows.
+			var foundIt = []sqlair.M{}
+			err = tx.Query(ctx, targetExistsStmt, m).GetAll(&foundIt)
+			if err != nil {
+				return errors.Annotatef(err, "verifying %q target exists", spec.Target.Key)
+			}
+			if len(foundIt) != 1 {
+				return errors.Annotatef(err, "permission target %q does not exist", spec.Target.Key)
+			}
+		}
+
+		// No IsErrConstraintForeignKey should be seen as both foreign keys
+		// have been checked.
+		err = tx.Query(ctx, insertPermissionStmt, m).Run()
+		if databaseutils.IsErrConstraintUnique(err) {
+			return errors.Annotatef(permissionerrors.AlreadyExists, "%q on %q", spec.User, spec.Target.Key)
+		} else if err != nil {
+			return errors.Annotatef(err, "adding permission %q for %q on %q", spec.Access, spec.User, spec.Target.Key)
+		}
+		return nil
+	})
+	if err != nil {
+		return corepermission.UserAccess{}, errors.Trace(err)
```

Comment:
```suggestion
		return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
```

We need to ensure that we don't leak sqlite/dqlite errors here.